Content.Shared/Construction/ConstructionConditions/WallmountCondition.cs

```diff
@@ -0,0 +1,72 @@
+﻿#nullable enable
+using Content.Shared.GameObjects.Components;
+using Content.Shared.Maps;
+using Content.Shared.Utility;
+using Robust.Shared.Interfaces.GameObjects;
+using Robust.Shared.IoC;
+using Robust.Shared.Map;
+using Robust.Shared.Maths;
+using Robust.Shared.Serialization;
+using System.Linq;
+
+namespace Content.Shared.Construction.ConstructionConditions
+{
+    public class WallmountCondition : IConstructionCondition
+    {
+        public void ExposeData(ObjectSerializer serializer) { }
+
+        public bool Condition(IEntity user, EntityCoordinates location, Direction direction)
+        {
+            // check that player doesn't try to build object facing wall other side
+            bool invalidDirection = FacingWallOtherSide(user, location, direction);
+            if (invalidDirection)
+                return false;
+
+            if (direction != Direction.North)
+            {
+                // check that building inside wall
+                var wallEnity = location.GetEntitiesInTile(true)
+                    .FirstOrDefault((e) => e.HasComponent<SharedWallComponent>());
```

Comment:
I'm really not a fan of adding SharedWall and think you should really just raycast instead probably, or just check collision properties, anything else.

When we get a bigger map (the end goal is roughly 8x - 10x the size of saltern) you're talking about potentially 10,000 - 20,000 dummy components just for this and pjb would probably shank me if I merged it.

Plus it doesn't handle other tile-sized blockers like vending machines, medical scanners, AME parts, etc.

Reply:
This is why we need tags

Reply:
NPCs definitely need tags for certain things at some point but in this case couldn't you just raycast to the nearest edge, so if you're building on the farside it'll block on the wall but on the nearside it won't for example? Or am I underthinking it.

Reply:
I need someway to check that this player trying build light on the wall. To distinguish wall from low wall or window, I need tag. Same logic used for low walls in [SharedCanBuildWindowOnTopComponent](https://github.com/space-wizards/space-station-14/blob/745401a41e219b88d2e71983c8f6b3a1aaa486e1/Content.Shared/Construction/ConstructionConditions/LowWallInTile.cs#L23)

So I'm not sure how raycast can fix that. Does SS14 some kind of raycast groups or layers? Should I add all walls to this group?

Reply:
Walls have a specific collision mask that other things don't have i.e. Opaque and MobImpassable together (not many things have Opaque) so you could probably check on that.

The raycast also means it'll handle building to the specific edge as well.

Reply:
I'm afraid it won't work. There are about 146 prototypes that are Opaque and MobImpassable. Furniture, airlocks and so on. 
Should I add new collision group for walls?

Reply:
What about walls 5 layers?